### PR TITLE
Implement "Create a Trusted Type" algorithm

### PR DESCRIPTION
**Pull request by TimvdLippe**
This algorithm is quite straightforward written in the specification,
but leads to some type awkwardness in Rust. Most notably, the callbacks
have different types and cannot be unified easily. They also return
different string types. Similarly, the returning objects are all unique
types and don't have a common denominator.

Therefore, rather than implementing it in 1-to-1 fashion with the
specification text, it instead uses callbacks to instruct the type
system of what to call when.

This is further complicated by the fact that the callback can exist
or not, as well as return a value or not. This requires multiple
unwrangling, combined with the fact that the algorithm should throw
or not.

All in all, the number of lines is relatively low compared to the
specification algorithm and the Rust compiler does a lot of heavy
lifting figuring out which type is what.

Part of https://github.com/servo/servo/issues/36258

### CODE REVIEW

**Review by jdm — CHANGES_REQUESTED** [review]
Nice solution!

File: components/script/dom/trustedtypepolicy.rs [hunk 1]
```diff
@@ -42,36 +150,87 @@ impl TrustedTypePolicyMethods<crate::DomTypeHolder> for TrustedTypePolicy {
...
+        arguments: Vec<HandleValue>,
         can_gc: CanGc,
-    ) -> DomRoot<TrustedHTML> {
-        // TODO(36258): handle arguments
-        TrustedHTML::new(data.to_string(), &self.global(), can_gc)
+    ) -> Fallible<DomRoot<TrustedHTML>> {
+        self.create_trusted_type(
+            || {
+                self.create_html.clone().map(|callback| {
+                    rooted!(in(*cx) let this_object = unsafe {JS_NewObject(*cx, ptr::null()) });
```
[the remark this answers is not in the document]
Comment (TimvdLippe) [reply components/script/dom/trustedtypepolicy.rs:164]: I tried to do this, but the code as you wrote it introduces a panic:

<details>

```
  │ assertion failed: !obj.is_null() (thread Script(1,1), at Projects/servo/components/script_bindings/callback.rs:240)
  │    0: servoshell::backtrace::print
  │              at Projects/servo/ports/servoshell/backtrace.rs:18:5
  │    1: servoshell::panic_hook::panic_hook
  │              at Projects/servo/ports/servoshell/panic_hook.rs:40:17
  │    2: core::ops::function::Fn::call
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79:5
  │    3: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/alloc/src/boxed.rs:2007:9
  │       std::panicking::rust_panic_with_hook
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/panicking.rs:836:13
  │    4: std::panicking::begin_panic_handler::{{closure}}
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/panicking.rs:694:13
  │    5: std::sys::backtrace::__rust_end_short_backtrace
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/sys/backtrace.rs:168:18
  │    6: rust_begin_unwind
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/panicking.rs:692:5
  │    7: core::panicking::panic_fmt
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/panicking.rs:75:14
  │    8: core::panicking::panic
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/panicking.rs:145:5
  │    9: script_bindings::callback::wrap_call_this_value
  │              at Projects/servo/components/script_bindings/callback.rs:240:5
  │   10: script_bindings::codegen::GenericBindings::TrustedTypePolicyFactoryBinding::CreateScriptCallback<D>::Call_
  │              at Projects/servo/target/debug/build/script_bindings-e177716eaf3b9e8a/out/Bindings/TrustedTypePolicyFactoryBinding.rs:273:27
  │   11: <script::dom::trustedtypepolicy::TrustedTypePolicy as script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::TrustedTypePolicyMethods<script::dom::bindings::codegen::DomTypeHolder::DomTypeHolder>>::CreateScript::{{closure}}::{{closure}}
  │              at Projects/servo/components/script/dom/trustedtypepolicy.rs:194:21
  │   12: core::option::Option<T>::map
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1119:29
  │   13: <script::dom::trustedtypepolicy::TrustedTypePolicy as script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::TrustedTypePolicyMethods<script::dom::bindings::codegen::DomTypeHolder::DomTypeHolder>>::CreateScript::{{closure}}
  │              at Projects/servo/components/script/dom/trustedtypepolicy.rs:189:17
  │   14: script::dom::trustedtypepolicy::TrustedTypePolicy::get_trusted_type_policy_value
  │              at Projects/servo/components/script/dom/trustedtypepolicy.rs:85:24
  │   15: script::dom::trustedtypepolicy::TrustedTypePolicy::create_trusted_type
  │              at Projects/servo/components/script/dom/trustedtypepolicy.rs:128:28
  │   16: <script::dom::trustedtypepolicy::TrustedTypePolicy as script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::TrustedTypePolicyMethods<script::dom::bindings::codegen::DomTypeHolder::DomTypeHolder>>::CreateScript
  │              at Projects/servo/components/script/dom/trustedtypepolicy.rs:187:9
  │   17: script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::createScript::{{closure}}::{{closure}}
  │              at Projects/servo/target/debug/build/script_bindings-e177716eaf3b9e8a/out/Bindings/TrustedTypePolicyBinding.rs:162:64
  │   18: script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::createScript::{{closure}}
  │              at Projects/servo/target/debug/build/script_bindings-e177716eaf3b9e8a/out/Bindings/TrustedTypePolicyBinding.rs:133:33
  │   19: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:305:13
  │   20: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:272:9
  │   21: std::panicking::try::do_call
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:584:40
  │   22: __rust_try
  │   23: std::panicking::try
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:547:19
  │       std::panic::catch_unwind
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:358:14
  │   24: mozjs::panic::wrap_panic
  │              at .cargo/git/checkouts/mozjs-dd2a25693c61d87d/e4d4f9a/mozjs/src/panic.rs:22:11
  │   25: script_bindings::codegen::GenericBindings::TrustedTypePolicyBinding::TrustedTypePolicy_Binding::createScript
  │              at Projects/servo/target/debug/build/script_bindings-e177716eaf3b9e8a/out/Bindings/TrustedTypePolicyBinding.rs:133:5
  │   26: CallJitMethodOp
  │              at /home/runner/work/mozjs/mozjs/mozjs-sys/src/jsglue.cpp:581:22
  │   27: script_bindings::utils::generic_call
  │              at Projects/servo/components/script_bindings/utils.rs:400:5
  │   28: script_bindings::utils::generic_method
  │              at Projects/servo/components/script_bindings/utils.rs:420:5
  │   29: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  │   30: _ZN2js9InterpretEP9JSContextRNS_8RunStateE
  │   31: _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   32: _ZN2js7ExecuteEP9JSContextN2JS6HandleIP8JSScriptEENS3_IP8JSObjectEENS2_13MutableHandleINS2_5ValueEEE
  │   33: _Z16JS_ExecuteScriptP9JSContextN2JS6HandleIP8JSScriptEENS1_13MutableHandleINS1_5ValueEEE
  │   34: mozjs::rust::wrappers::JS_ExecuteScript
  │              at .cargo/git/checkouts/mozjs-dd2a25693c61d87d/e4d4f9a/mozjs/src/rust.rs:1135:51
  │   35: script::dom::globalscope::GlobalScope::evaluate_script_on_global_with_result
  │              at Projects/servo/components/script/dom/globalscope.rs:2611:26
  │   36: script::dom::htmlscriptelement::HTMLScriptElement::run_a_classic_script
  │              at Projects/servo/components/script/dom/htmlscriptelement.rs:1077:9
  │   37: script::dom::htmlscriptelement::HTMLScriptElement::execute
  │              at Projects/servo/components/script/dom/htmlscriptelement.rs:1039:17
  │   38: script::dom::htmlscriptelement::HTMLScriptElement::prepare
  │              at Projects/servo/components/script/dom/htmlscriptelement.rs:913:25
  │   39: script::dom::servoparser::ServoParser::tokenize
  │              at Projects/servo/components/script/dom/servoparser/mod.rs:659:13
  │   40: script::dom::servoparser::ServoParser::parse_sync
  │              at Projects/servo/components/script/dom/servoparser/mod.rs:587:9
  │   41: script::dom::servoparser::ServoParser::resume_with_pending_parsing_blocking_script
  │              at Projects/servo/components/script/dom/servoparser/mod.rs:349:13
  │   42: script::dom::document::Document::process_pending_parsing_blocking_script
  │              at Projects/servo/components/script/dom/document.rs:2906:13
  │   43: script::dom::document::Document::pending_parsing_blocking_script_loaded
  │              at Projects/servo/components/script/dom/document.rs:2892:9
  │   44: script::dom::htmlscriptelement::finish_fetching_a_classic_script
  │              at Projects/servo/components/script/dom/htmlscriptelement.rs:368:13
  │   45: <script::dom::htmlscriptelement::ClassicContext as net_traits::FetchResponseListener>::process_response_eof
  │              at Projects/servo/components/script/dom/htmlscriptelement.rs:518:9
  │   46: <net_traits::FetchResponseMsg as net_traits::Action<T>>::process
  │              at Projects/servo/components/shared/net/lib.rs:356:25
  │   47: <script::network_listener::ListenerTask<A,Listener> as script::task::TaskOnce>::run_once
  │              at Projects/servo/components/script/network_listener.rs:118:13
  │   48: <script::task::CancellableTask<T> as script::task::TaskOnce>::run_once
  │              at Projects/servo/components/script/task.rs:106:13
  │   49: <T as script::task::TaskBox>::run_box
  │              at Projects/servo/components/script/task.rs:57:9
  │   50: script::script_thread::ScriptThread::handle_msg_from_script
  │              at Projects/servo/components/script/script_thread.rs:1972:17
  │   51: script::script_thread::ScriptThread::handle_msgs::{{closure}}
  │              at Projects/servo/components/script/script_thread.rs:1509:60
  │   52: script::script_thread::ScriptThread::profile_event
  │              at Projects/servo/components/script/script_thread.rs:1721:13
  │   53: script::script_thread::ScriptThread::handle_msgs
  │              at Projects/servo/components/script/script_thread.rs:1500:27
  │   54: script::script_thread::ScriptThread::start
  │              at Projects/servo/components/script/script_thread.rs:978:15
  │   55: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}::{{closure}}
  │              at Projects/servo/components/script/script_thread.rs:431:25
  │   56: profile_traits::mem::ProfilerChan::run_with_memory_reporting
  │              at Projects/servo/components/shared/profile/mem.rs:91:9
  │   57: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}
  │              at Projects/servo/components/script/script_thread.rs:429:17
  │   58: std::sys::backtrace::__rust_begin_short_backtrace
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:152:18
  │   59: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:564:17
  │   60: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:272:9
  │   61: std::panicking::try::do_call
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:584:40
  │   62: __rust_try
  │   63: std::panicking::try
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:547:19
  │       std::panic::catch_unwind
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:358:14
  │       std::thread::Builder::spawn_unchecked_::{{closure}}
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:562:30
  │   64: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │              at .rustup/toolchains/1.85.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  │   65: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/alloc/src/boxed.rs:1993:9
  │       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/alloc/src/boxed.rs:1993:9
  │       std::sys::pal::unix::thread::Thread::new::thread_start
  │              at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/sys/pal/unix/thread.rs:106:17
```

</details>

I have also tried the following code with the same result:

```
rooted!(in(*cx) let this_object: *mut JSObject = ptr::null_mut());
```

I see no other usages of `.Call_` that pass in a null pointer. Is that possible to write with the current implementation of `wrap_call_this_value`?

[the remark this answers is not in the document]
Comment (gterzian) [reply components/script/dom/trustedtypepolicy.rs:164]: Actually it may also be `SafeHandleValue::null()` because I can imagine that `Call__` wants to set `this` to undefined, and here as per the spec it should be null. To be tried out...

[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/trustedtypepolicy.rs:164]: Using an undefined value instead of null is an observable difference from the spec, unfortunately.

[the remark this answers is not in the document]
Comment (TimvdLippe) [reply components/script/dom/trustedtypepolicy.rs]: Fixed by updating `wrap_call_this_value` to handle `null`. Also added a WPT test as apparently this wasn't covered yet in any of the existing tests (hence I missed the `{}` vs `null`)

**Review by jdm — APPROVED** [review]
Thanks for the test!

**Comment by TimvdLippe** [comment]
Not sure why the DCO check isn't firing, blocking the merge?
